tests/returntypes/gradients/test_hadamard_grad.py

```diff
@@ -0,0 +1,853 @@
+
+
+def cost3(x):
+    """Cost function."""
+    qml.Rot(*x, wires=0)
+    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+
+def cost4(x):
+    """Cost function."""
+    qml.Rot(*x, wires=0)
+    return qml.probs([0, 1])
+
+
+def cost5(x):
+    """Cost function."""
+    qml.Rot(*x, wires=0)
+    return [qml.probs([0, 1])]
+
+
+def cost6(x):
+    """Cost function."""
+    qml.Rot(*x, wires=0)
+    return [qml.probs([0, 1]), qml.probs([2, 3])]
+
+
+class TestHadamardGrad:
+    """Unit tests for the hadamard_grad function"""
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, 2 * np.pi, 7))
+    @pytest.mark.parametrize("G", [qml.RX, qml.RY, qml.RZ, qml.PhaseShift, qml.U1])
+    def test_pauli_rotation_gradient(self, G, theta, tol):
+        """Tests that the automatic gradients of Pauli rotations are correct."""
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.QubitStateVector(np.array([1.0, -1.0], requires_grad=False) / np.sqrt(2), wires=0)
+            G(theta, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        tape.trainable_params = {1}
+
+        tapes, fn = qml.gradients.hadamard_grad(tape)
+
+        if isinstance(G(theta, wires=[0]), (qml.RX, qml.RY, qml.RZ)):
+            assert len(tapes) == 1
+        else:
+            assert len(tapes) == 2
+
+        hadamard_val = fn(dev.batch_execute(tapes))
+
+        assert isinstance(hadamard_val, np.ndarray)
+        assert hadamard_val.shape == ()
+
+        tapes, fn = qml.gradients.param_shift(tape)
+        param_shift_val = fn(dev.batch_execute(tapes))
+
+        assert np.allclose(hadamard_val, param_shift_val, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, 2 * np.pi, 7))
+    def test_rot_gradient(self, theta, tol):
+
+        """Tests that the automatic gradient of an arbitrary Euler-angle-parameterized gate is correct."""
+        dev = qml.device("default.qubit", wires=2)
+        params = np.array([theta, theta**3, np.sqrt(2) * theta])
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.QubitStateVector(np.array([1.0, -1.0], requires_grad=False) / np.sqrt(2), wires=0)
+            qml.Rot(*params, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        tape.trainable_params = {1, 2, 3}
+
+        tapes, fn = qml.gradients.hadamard_grad(tape)
+        num_params = len(tape.trainable_params)
+        assert len(tapes) == num_params
+
+        hadamard_val = fn(dev.batch_execute(tapes))
+
+        assert isinstance(hadamard_val, tuple)
+        assert len(hadamard_val) == num_params
+
+        tapes, fn = qml.gradients.param_shift(tape)
+        param_val = fn(dev.batch_execute(tapes))
+
+        for a_val, n_val in zip(hadamard_val, param_val):
+            assert np.allclose(a_val, n_val, atol=tol, rtol=0)
```

Comment:
Doesn't this work with a single `np.allclose` call?
```suggestion
        assert np.allclose(hadamard_val, param_val, atol=tol, rtol=0)
```